File: backend/Services/Parsers/ContextoParser.cs
using System.Text.RegularExpressions;
using backend.Models;

namespace backend.Services.Parsers
{
    public class ContextoParser : ResultParser
    {
        private ILogger<ContextoParser> _logger;

        public ContextoParser(ILogger<ContextoParser> logger) : base(logger)
        {
            _logger = logger;
        }

        public override bool CountWinner => false;
        public override bool Default => false;
        public override string GameName => "Contexto";
        public override bool GolfScoring => true;
        public override string? HelpText => null;
        private const string CompletedGroup = "completed";
        private const string ScoreGroup = "score";
        private const string HintGroup = "hint";
        protected override Regex Parser => new Regex($@"^I played contexto.me #\d+ (?<{CompletedGroup}>[^\d]+)(?<{ScoreGroup}>\d+) guesses( and (?<{HintGroup}>\d+) hints)?");
        public override string Url => "https://contexto.me/";

        protected override string GetCleanResult(string result, Match parserResults)
        {
            return result.Trim();
        }

        public override string? GetScoreValue(DailyResult dailyResult)
        {
            if (dailyResult.Scores == null)
            {
                return null;
            }

            if (dailyResult.Scores.Count == 1)
            {
                return dailyResult.Scores.Single().ToString();
            }

            return "\"" + string.Join(",", dailyResult.Scores
                .Select((i) => dailyResult.Scores[i].ToString())) + "\"";
        }

        protected override DailyResult SetScore(DailyResult dailyResult, Match parserResults)
        {
            if (!parserResults.Groups.ContainsKey(CompletedGroup) ||
                !parserResults.Groups.ContainsKey(ScoreGroup) ||
                !Int32.TryParse(parserResults.Groups[ScoreGroup].Value, out var score) ||
                // gave up, no score
                parserResults.Groups[CompletedGroup].Value.StartsWith("but I gave up in"))
            {
                return dailyResult;
            }

            var scores = new List<int>();
            if (parserResults.Groups.ContainsKey(HintGroup) && Int32.TryParse(parserResults.Groups[HintGroup].Value, out var hints))
            {
                scores.Add(hints);
            }

            scores.Add(score);

            dailyResult.Scores = scores;

            return dailyResult;
        }
    }
}

File: lambdas/src/parsers/contexto-parser.ts
import { DailyResult } from '../models/daily-result';
import { ResultParser, getGroup, hasGroup } from './base/result-parser';

export class ContextoParser extends ResultParser {
  readonly countWinner = true;
  readonly gameName = 'Contexto';
  readonly golfScoring = true;
  readonly helpText = null;
  readonly url = 'https://contexto.me/';
  
  private static readonly COMPLETED_GROUP = 'completed';
  private static readonly SCORE_GROUP = 'score';
  private static readonly HINT_GROUP = 'hint';
  
  // Matches "I played contexto.me #822 and got it in 11 guesses" or with hints
  protected readonly parser = new RegExp(
    `^I played contexto\\.me #\\d+ (?<${ContextoParser.COMPLETED_GROUP}>[^\\d]+)(?<${ContextoParser.SCORE_GROUP}>\\d+) guesses( and (?<${ContextoParser.HINT_GROUP}>\\d+) hints)?`
  );
  
  protected getCleanResult(result: string, match: RegExpMatchArray): string {
    return result.trim();
  }
  
  getScoreValue(dailyResult: DailyResult): string | null {
    if (!dailyResult.scores) return null;
    
    if (dailyResult.scores.length === 1) {
      return dailyResult.scores[0].toString();
    }
    
    return '"' + dailyResult.scores.join(',') + '"';
  }
  
  protected setScore(dailyResult: DailyResult, match: RegExpMatchArray): DailyResult {
    const completedStr = getGroup(match, ContextoParser.COMPLETED_GROUP);
    const scoreStr = getGroup(match, ContextoParser.SCORE_GROUP);
    
    // gave up, no score
    if (!completedStr || !scoreStr || completedStr.startsWith('but I gave up in')) {
      return dailyResult;
    }
    
    const score = parseInt(scoreStr, 10);
    if (isNaN(score)) return dailyResult;
    
    const scores: number[] = [];
    const hintStr = getGroup(match, ContextoParser.HINT_GROUP);
    if (hintStr) {
      const hints = parseInt(hintStr, 10);
      if (!isNaN(hints)) {
        scores.push(hints);
      }
    }
    
    scores.push(score);
    dailyResult.scores = scores;
    
    return dailyResult;
  }
}
